connector/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryDataWriterHelper.java

```diff
@@ -0,0 +1,364 @@
+                  messageSize, HARD_LIMIT_APPEND_ROWS_REQUEST_SIZE));
+        }
+        protoRows.addSerializedRows(message);
+      }
+      sendAppendRowsRequest();
+    }
+
+    protoRows.addSerializedRows(message);
+    appendRequestSizeBytes += messageSize;
+    appendRequestRowCount++;
+  }
+
+  /**
+   * Sends an AppendRowsRequest to the BigQuery Storage Write API.
+   *
+   * @throws IOException If the append rows request fails: either by returning the wrong offset
+   *     (deduplication error) or if the response contains an error.
+   */
+  private void sendAppendRowsRequest() throws IOException {
+    long offset = writeStreamRowCount;
+
+    Storage.AppendRowsRequest appendRowsRequest = createAppendRowsRequest(offset);
+
+    ApiFuture<Storage.AppendRowsResponse> appendRowsResponseApiFuture =
+        streamWriter.append(appendRowsRequest);
+    validateAppendRowsResponse(appendRowsResponseApiFuture, offset);
+
+    clearProtoRows();
+    this.writeStreamRowCount += appendRequestRowCount;
+    this.writeStreamSizeBytes += appendRequestSizeBytes;
+    this.appendRequestRowCount = 0;
+    this.appendRequestSizeBytes = 0;
+  }
+
+  /**
+   * Helper method to create a BigQuery Storage Write API AppendRowsRequest.
+   *
+   * @param offset The offset to be used in the request: equal to how many rows have already been
+   *     appended.
+   * @return The AppendRowsRequest
+   * @see com.google.cloud.bigquery.storage.v1alpha2.Storage.AppendRowsRequest
+   */
+  private Storage.AppendRowsRequest createAppendRowsRequest(long offset) {
+    Storage.AppendRowsRequest.Builder requestBuilder =
+        Storage.AppendRowsRequest.newBuilder().setOffset(Int64Value.of(offset));
+
+    Storage.AppendRowsRequest.ProtoData.Builder dataBuilder =
+        Storage.AppendRowsRequest.ProtoData.newBuilder();
+    dataBuilder.setWriterSchema(protoSchema);
+    dataBuilder.setRows(protoRows.build());
+
+    requestBuilder.setProtoRows(dataBuilder.build()).setWriteStream(writeStreamName);
+
+    return requestBuilder.build();
+  }
+
+  /**
+   * Validates an AppendRowsResponse, after retrieving its future: makes sure the responses' future
+   * matches the expectedOffset, and returned with no errors.
+   *
+   * @param appendRowsResponseApiFuture The future of the AppendRowsResponse
+   * @param expectedOffset The expected offset to be returned by the response.
+   * @throws IOException If the response returned with error, or the offset did not match the
+   *     expected offset.
+   */
+  private void validateAppendRowsResponse(
+      ApiFuture<Storage.AppendRowsResponse> appendRowsResponseApiFuture, long expectedOffset)
+      throws IOException {
+    Storage.AppendRowsResponse appendRowsResponse = null;
+    try {
+      appendRowsResponse = appendRowsResponseApiFuture.get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException("Could not retrieve AppendRowsResponse.", e);
```

Comment:
Please change RuntimeException to a subclass with better description
nit: remove dot in the end of the message